src/hotspot/share/gc/g1/g1PageBasedVirtualSpace.cpp

```diff
@@ -250,3 +251,49 @@ void G1PageBasedVirtualSpace::print() {
   print_on(tty);
 }
 #endif
+
+void G1PageBasedVirtualSpace::persist_for_checkpoint() {
+  crac::MemoryPersister persister;
+  bool flip = _committed.at(0);
+  size_t index = 0;
+  while (index < _committed.size()) {
+    size_t next;
+    if (flip) {
+      next = _committed.find_first_clear_bit(index);
+      size_t length = (next - index) * _page_size;
+      if (!persister.store((char *) _low_boundary + index * _page_size, length, length)) {
+        fatal("Failed to persist committed virtual space node range");
+      }
+    } else {
+      next = _committed.find_first_set_bit(index);
+      if (!persister.store_gap((char *) _low_boundary + index * _page_size, (next - index) * _page_size)) {
+        fatal("Failed to persist uncommitted virtual space node range");
+      }
+    }
+    flip = !flip;
+    index = next;
+  }
+}
+
+void G1PageBasedVirtualSpace::load_on_restore() {
+  crac::MemoryLoader loader;
+  bool flip = _committed.at(0);
+  size_t index = 0;
+  while (index < _committed.size()) {
+    size_t next;
+    if (flip) {
+      next = _committed.find_first_clear_bit(index);
+      size_t length = (next - index) * _page_size;
+      if (!loader.load((char *) _low_boundary + index * _page_size, length, length, false)) {
+        fatal("Cannot restore committed virtual space node");
+      }
+    } else {
+      next = _committed.find_first_set_bit(index);
+      if (!loader.load_gap((char *) _low_boundary + index * _page_size, (next - index) * _page_size)) {
+        fatal("Cannot map uncommitted virtual space node");
+      }
+    }
+    flip = !flip;
+    index = next;
+  }
```

Comment:
Regarding the quote, I mean it could be:
```
// _snapshot_committed - copy of _committed only to be verified during checkpoint 

void G1PageBasedVirtualSpace::persist_for_checkpoint() {
  memcpy(_snapshot_committed, _committed, , sizeof(_snapshot_committed));
  ... // the rest of the current impl
}
 
void G1PageBasedVirtualSpace::load_on_restore() {
   guarantee(0 == memcmp(_snapshot_committed, _committed, sizeof(_snapshot_committed),
       "unexpected commit during checkpoint");
 }
 ```
 
And that's it -- G1PageBasedVirtualSpace just checks own invariants, an relies on MemoryPersister/Loader to load all necessary pages. This looks like as safe as the current implementation, and it could be a bit faster (no need to decide what do load here, and search that  in the index, just let Loader load everything), but the aim is to make code simpler and avoid repetitions.

Reply:
We would need to allocate a new memory, get into impl details of the bitmap (`_committed._map`) etc.; I wouldn't do it in release mode for sure. Given the complexity the check seems as an overkill.

(for the record, I have no issues with restoring the memory just from the info in MemoryLoader)

Reply:
> I have no issues with restoring the memory just from the info in MemoryLoader

 :+1: 
 
 > We would need to allocate a new memory, get into impl details of the bitmap (`_committed._map`) etc.; I wouldn't do it in release mode for sure. Given the complexity the check seems as an overkill.

Hmm, that indeed does not look nice. Thinking about alternatives, we may have a boolean flag set to true during checkpoints only, and check the flag during operations with _committed, asserting accidental change. I agree that the checks should be pretty cheap in production and easy to implement.